edb/schema/referencing.py

```diff
@@ -245,47 +279,66 @@ def _classname_from_ast(cls, schema, astnode, context):
             pnn = sn.get_specialized_name(base_name, referrer_name, *quals)
             name = sn.Name(name=pnn, module=referrer_name.module)
 
+        assert isinstance(name, sn.Name)
         return name
 
     @classmethod
     def _classname_from_name(
         cls,
         name: sn.SchemaName,
-        referrer_name: str,
+        referrer_name: sn.SchemaName,
     ) -> sn.Name:
         base_name = sn.shortname_from_fullname(name)
         quals = cls._classname_quals_from_name(name)
         pnn = sn.get_specialized_name(base_name, referrer_name, *quals)
         return sn.Name(name=pnn, module=referrer_name.module)
 
     @classmethod
-    def _classname_quals_from_ast(cls, schema, astnode, base_name,
-                                  referrer_name, context):
+    def _classname_quals_from_ast(cls,
+                                  schema: s_schema.Schema,
+                                  astnode: qlast.NamedDDL,
+                                  base_name: sn.SchemaName,
+                                  referrer_name: str,
+                                  context: sd.CommandContext
+                                  ) -> Tuple[str, ...]:
         return ()
 
     @classmethod
-    def _classname_quals_from_name(cls, name: sn.SchemaName) -> Tuple[str]:
+    def _classname_quals_from_name(cls,
+                                   name: sn.SchemaName
+                                   ) -> Tuple[str, ...]:
         return ()
 
     @classmethod
-    def _name_qual_from_exprs(cls, schema, exprs):
+    def _name_qual_from_exprs(cls,
+                              schema: s_schema.Schema,
+                              exprs: Iterable[str]) -> str:
         m = hashlib.sha1()
         for expr in exprs:
             m.update(expr.encode())
         return m.hexdigest()
 
-    def _get_ast_node(self, schema, context):
+    def _get_ast_node(self,
+                      schema: s_schema.Schema,
+                      context: sd.CommandContext
+                      ) -> Type[qlast.DDLOperation]:
         subject_ctx = self.get_referrer_context(context)
         ref_astnode = getattr(self, 'referenced_astnode', None)
         if subject_ctx is not None and ref_astnode is not None:
-            return ref_astnode
+            # type ignore below, because asserting issubclass doesn't work
+            # in the current stable version of mypy
+            return ref_astnode  # type: ignore
```

Comment:
My understanding is that:

```
assert issubclass(ref_astnode, qlast.DDLOperation)
```

Should make mypy happy. Because if you assert that _A_ is a subclass of _B_, you assert that A is a type of B ("Type[B]").

Instead, mypy still considers `ref_astnode` to be `Any`.

---
Another solution would be to do:

```python
        ref_astnode: Type[qlast.DDLOperation] = getattr(self,
                                                        'referenced_astnode',
                                                        None)
        if subject_ctx is not None and ref_astnode is not None:
            return ref_astnode
```

But it doesn't give any concrete value; since `getattr` can return anything.